assets/js/hooks/session/cursor_history.js

```diff
@@ -0,0 +1,136 @@
+/**
+ * Allows for recording a sequence of focused cells with the focused line
+ * and navigate inside this stack.
+ */
+export default class CursorHistory {
+  /** @private */
+  entries = [];
+
+  /** @private */
+  index = -1;
+
+  /**
+   * Adds a new cell to the stack.
+   *
+   * If the stack length is greater than the stack limit,
+   * it will remove the oldest entries.
+   */
+  push(cellId, line, offset) {
+    const entry = { cellId, line, offset };
+
+    if (this.isSameCell(cellId)) {
+      this.entries[this.index] = entry;
+    } else {
+      if (this.entries[this.index + 1] !== undefined) {
+        this.entries = this.entries.slice(0, this.index + 1);
+      }
+
+      this.entries.push(entry);
+      this.index++;
+    }
+
+    if (this.entries.length > 20) {
+      this.entries.shift();
+      this.index--;
+    }
+  }
+
+  /**
+   * Removes all matching cells with given id from the stack.
+   */
+  removeAllFromCell(cellId) {
+    // We need to make sure the last entry from history
+    // doesn't belong to the given cell id that we need
+    // to remove from the entries list.
+    let cellIdCount = 0;
+
+    for (let i = 0; i <= this.index; i++) {
+      const entry = this.get(i);
+      if (entry === null) continue;
+
+      if (entry.cellId === cellId) cellIdCount++;
+    }
+
+    this.index = this.index - cellIdCount;
+    if (this.index < 0) this.index = 0;
+
+    this.entries = this.entries.filter((entry) => entry.cellId !== cellId);
```

Comment:
We only want to set 0 if there are any cells left

```suggestion
    this.entries = this.entries.filter((entry) => entry.cellId !== cellId);
    this.index = this.index - cellIdCount;
    if (this.index === -1 && this.entries.length > 0) this.index = 0;
```